File: app/src/main/java/com/example/pawel/weatherapp/project/ColorHelper.java
package com.example.pawel.weatherapp.project;

import android.graphics.Color;

public class ColorHelper {

    static String TAG = "ColorHelper";
    private static int MAX_TEMP = 40;
    private static int MIN_TEMP = -20;
    private static int TEMP_DIFF = MAX_TEMP - MIN_TEMP;
    private static int HOT_COLOR = Color.parseColor("#FF8000");
    static int[] START_COLOR = colorToBasic(HOT_COLOR);
    private static int COLD_COLOR = Color.parseColor("#165B88");
    static int[] END_COLOR = colorToBasic(COLD_COLOR);

    private static int[] colorToBasic(int color) {

        int alpha = Color.alpha(color);
        int red = Color.red(color);
        int green = Color.green(color);
        int blue = Color.blue(color);
        return new int[]{alpha, red, green, blue};
    }

    public static int pickHex(float progress) {
        return pickHex(START_COLOR, END_COLOR, progress, 1 - progress);
    }

    public static int pickHex(int[] color1, int[] color2, float weight) {
        return pickHex(color1, color2, weight, 1 - weight);
    }

    public static int pickHex(int[] color1, int[] color2, float weight, float restWeight) {
        return Color.argb(Math.round(color2[0] * weight + color1[0] * restWeight),
                Math.round(color2[1] * weight + color1[1] * restWeight),
                Math.round(color2[2] * weight + color1[2] * restWeight),
                Math.round(color2[3] * weight + color1[3] * restWeight));
    }

    public static int colorForTemperature(float temp) {

        if (temp > MAX_TEMP)
            return HOT_COLOR;
        else if (temp < MIN_TEMP)
            return COLD_COLOR;
        else {
            float tempPercent = 1 - Math.abs(temp / TEMP_DIFF);
            return pickHex(START_COLOR, END_COLOR, tempPercent);
        }
    }

    public static int getThumbColor(int[] tempArray) {
        return getThumbColor(tempArray, 0);
    }

    public static int getThumbColor(int[] tempArray, int progress) {
        return shadeColor(getTempColor(tempArray, progress));
    }

    public static int getTempColor(int[] tempArray) {
        return Color.argb(tempArray[0], tempArray[1], tempArray[2], tempArray[3]);
    }

    public static int getTempColor(int[] tempArray, int progress) {

        if (progress < 25)
            return pickHex(colorToBasic(tempArray[0]), colorToBasic(tempArray[1]), progress / 5f);
        else if (progress < 50)
            return pickHex(colorToBasic(tempArray[1]), colorToBasic(tempArray[2]), progress / 5f);
        else if (progress < 75)
            return pickHex(colorToBasic(tempArray[2]), colorToBasic(tempArray[3]), progress / 5f);
        else
            return pickHex(colorToBasic(tempArray[3]), colorToBasic(tempArray[4]), progress / 5f);
    }

    private static int shadeColor(int color) {
        int[] base = colorToBasic(color);
        for (int i = 1; i < 4; i++) {
            base[i] += base[i] / 3;
        }
        return Color.argb(base[0], base[1], base[2], base[3]);
    }
}

File: weatherlibwithcityphotos/src/main/java/com/example/weatherlibwithcityphotos/ColorHelper.kt
package com.example.weatherlibwithcityphotos

import android.graphics.Color
import android.util.Log

class ColorHelper {

    companion object {

        private val MAX_TEMP = 40
        private val MIN_TEMP = -20
        private val TEMP_DIFF = MAX_TEMP - MIN_TEMP
        private val HOT_COLOR = Color.parseColor("#FF8000")
        private val COLD_COLOR = Color.parseColor("#165B88")
        val START_COLOR = colorToBasic(HOT_COLOR)
        val END_COLOR = colorToBasic(COLD_COLOR)


        private fun colorToBasic(color: Int): IntArray {

            val alpha = Color.alpha(color)
            val red = Color.red(color)
            val green = Color.green(color)
            val blue = Color.blue(color)

            return intArrayOf(alpha, red, green, blue)
        }

        fun pickHex(progress: Float): Int {
            return pickHex(START_COLOR, END_COLOR, progress, 1 - progress)
        }

        fun pickHex(color1: IntArray, color2: IntArray, weight: Float, restWeight: Float = 1 - weight): Int {
            Log.i("Color", "Progress : $weight rest : $restWeight")
            return Color.argb(Math.round(color2[0] * weight + color1[0] * restWeight),
                    Math.round(color2[1] * weight + color1[1] * restWeight),
                    Math.round(color2[2] * weight + color1[2] * restWeight),
                    Math.round(color2[3] * weight + color1[3] * restWeight))
        }

        fun colorForTemperature(temp: Float): Int {
            return when {
                temp > MAX_TEMP -> HOT_COLOR
                temp < MIN_TEMP -> COLD_COLOR
                else -> {
                    val tempPercent = 1 - Math.abs(temp / TEMP_DIFF)
                    pickHex(START_COLOR, END_COLOR, tempPercent)
                }
            }
        }

        fun getThumbColor(progress: Int, tempArray: IntArray): Int {
            return shadeColor(getTempColor(progress, tempArray))
        }

        fun getTempColor(progress: Int, tempArray: IntArray): Int {
            return when {
                progress < 25 -> {
                    pickHex(colorToBasic(tempArray[0]), colorToBasic(tempArray[1]), progress / 5f)
                }
                progress < 50 -> {
                    pickHex(colorToBasic(tempArray[1]), colorToBasic(tempArray[2]), progress / 5f)
                }
                progress < 75 -> {
                    pickHex(colorToBasic(tempArray[2]), colorToBasic(tempArray[3]), progress / 5f)
                }
                else -> {
                    pickHex(colorToBasic(tempArray[3]), colorToBasic(tempArray[4]), progress / 5f)
                }
            }
        }

        private fun shadeColor(color: Int): Int {
            val base = colorToBasic(color)
            for (i in 1 until base.size) {
                base[i] += base[i] * 1 / 3
            }
            return Color.argb(base[0], base[1], base[2], base[3])
        }
    }
}